src/marketmap.rs

```diff
@@ -0,0 +1,286 @@
+use std::cell::{Cell, RefCell};
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::{Arc, Mutex};
+
+use crate::event_emitter::EventEmitter;
+use crate::memcmp::get_market_filter;
+use crate::utils::{decode, get_ws_url};
+use crate::websocket_program_account_subscriber::{
+    ProgramAccountUpdate, WebsocketProgramAccountOptions, WebsocketProgramAccountSubscriber,
+};
+use crate::{DataAndSlot, SdkResult};
+use anchor_lang::AccountDeserialize;
+use dashmap::DashMap;
+use drift::state::oracle::OracleSource;
+use drift::state::perp_market::PerpMarket;
+use drift::state::spot_market::SpotMarket;
+use drift::state::user::MarketType;
+use serde_json::json;
+use solana_account_decoder::UiAccountEncoding;
+use solana_client::nonblocking::rpc_client::RpcClient;
+use solana_client::rpc_config::{RpcAccountInfoConfig, RpcProgramAccountsConfig};
+use solana_client::rpc_request::RpcRequest;
+use solana_client::rpc_response::{OptionalContext, RpcKeyedAccount};
+use solana_sdk::commitment_config::CommitmentConfig;
+use solana_sdk::pubkey::Pubkey;
+
+pub trait Market {
+    fn market_index(&self) -> u16;
+    fn market_type() -> MarketType;
+    fn oracle_info(&self) -> (Pubkey, OracleSource);
+}
+
+impl Market for PerpMarket {
+    fn market_index(&self) -> u16 {
+        self.market_index
+    }
+
+    fn market_type() -> MarketType {
+        MarketType::Perp
+    }
+
+    fn oracle_info(&self) -> (Pubkey, OracleSource) {
+        (self.amm.oracle, self.amm.oracle_source)
+    }
+}
+
+impl Market for SpotMarket {
+    fn market_index(&self) -> u16 {
+        self.market_index
+    }
+
+    fn market_type() -> MarketType {
+        MarketType::Spot
+    }
```

Comment:
can do a trait associated const for this
`const MARKET_TYPE: MarketType = MarketType::Spot;`